core/src/main/scala/org/apache/spark/scheduler/Task.scala

```diff
@@ -106,7 +106,21 @@ private[spark] abstract class Task[T](val stageId: Int, var partitionId: Int) ex
     if (interruptThread && taskThread != null) {
       taskThread.interrupt()
     }
-  }  
+  }
+
+  override def hashCode(): Int = {
+    31 * stageId.hashCode() + partitionId.hashCode()
+  }
+
+  def canEqual(other: Any): Boolean = other.isInstanceOf[Task[_]]
+
+  override def equals(other: Any): Boolean = other match {
+    case that: Task[_] =>
+      (that canEqual this) &&
```

Comment:
According to the logic of `DAGScheduler`, I think once the stage id of 2 tasks equals, the 2 tasks must be same type, so we don't need to check the type equality here.